numba/cpython/unicode.py

```diff
@@ -1644,24 +1644,49 @@ def unicode_getitem(s, idx):
         if isinstance(idx, types.Integer):
             def getitem_char(s, idx):
                 idx = normalize_str_idx(idx, len(s))
-                ret = _empty_string(s._kind, 1, s._is_ascii)
-                _set_code_point(ret, 0, _get_code_point(s, idx))
+                cp = _get_code_point(s, idx)
+                kind = _codepoint_to_kind(cp)
+                is_ascii = _codepoint_is_ascii(cp)
+                ret = _empty_string(kind, 1, is_ascii)
+                _set_code_point(ret, 0, cp)
                 return ret
             return getitem_char
         elif isinstance(idx, types.SliceType):
             def getitem_slice(s, idx):
                 slice_idx = _normalize_slice(idx, len(s))
                 span = _slice_span(slice_idx)
 
-                if slice_idx.step == 1:
+                cp = _get_code_point(s, slice_idx.start)
+                kind = _codepoint_to_kind(cp)
+                is_ascii = _codepoint_is_ascii(cp)
+
+                # Check slice to see if it's homogeneous in kind
+                for i in range(slice_idx.start + slice_idx.step,
+                               slice_idx.stop, slice_idx.step):
+                    cp = _get_code_point(s, i)
+                    is_ascii |= _codepoint_is_ascii(cp)
+                    new_kind = _codepoint_to_kind(cp)
+                    if kind != new_kind:
+                        kind = _pick_kind(kind, new_kind)
+                    # TODO: it might be possible to break here if the kind
+                    # is PY_UNICODE_4BYTE_KIND but there's potentially
```

Comment:
Perhaps it would be good to enhance this description with an additional sentence as to what the symptoms of breaking here might be?

Reply:
I'm not sure what this means? If the kind is already the widest kind then break, as looping to look for a bigger kind is redundant. Doing this would require that the strings are of the minimal kind required to hold their code point, but as noted in the comment, I suspect other internal functions aren't necessarily doing the correct thing.

Reply:
ah, so some internal stings have a kind larger than required to hold their code point?

Reply:
> ah, so some internal stings have a kind larger than required to hold their code point?

I suspect yes. This is on the basis that it hasn't been enforced.

Reply:
I think it would be worth adding a note about this insight. I left a suggestion below on how to make this comment a little clearer. Feel free to adopt that or suggest a better alternative.